common/namespace/registry.go

```diff
@@ -159,6 +169,14 @@ type (
 		callbackLock     sync.Mutex
 		prepareCallbacks map[any]PrepareCallbackFn
 		callbacks        map[any]CallbackFn
+
+		// State-change callbacks. To avoid races, callbacks are first added to
+		// newStateChangeCallbacks. The refresh loop will call those on all namespaces,
+		// then move them to stateChangeCallbacks, where they will be called only on
+		// namespaces that change state.
+		stateChangeCallbacksLock sync.Mutex
+		stateChangeCallbacks     map[any]StateChangeCallbackFn
+		newStateChangeCallbacks  map[any]StateChangeCallbackFn
```

Comment:
As we discussed, we want to remove stateChangeCallbacksLock and newStateChangeCallbacks. We are only going use one lock and one stateChangeCallbacks collection.